core/gather/gatherers/anchor-elements.js

```diff
@@ -38,10 +38,71 @@ function collectAnchorElements() {
     return onclick.slice(0, 1024);
   }
 
+  /** @param {HTMLElement|SVGElement|Text|ChildNode} node */
+  function getTrimmedInnerText(node) {
+    return node instanceof HTMLElement
+      ? node.innerText.trim()
+      : (node.textContent ? node.textContent.trim() : '');
+  }
+
+  /**
+   * @param {HTMLElement|SVGElement} node
+   * @param {string|null} currentLang
+   * @return {string}
+   */
+  function getLangOfInnerText(node, currentLang = null) {
+    if (currentLang === null) {
+      const parentWithLang = node.closest('[lang]');
+
+      // TODO: fallbacks to pragma-set-default-language or HTTP header
+      currentLang = !parentWithLang ? '' : parentWithLang.getAttribute('lang');
+    }
+
+    const innerElsWithLang = node.querySelectorAll('[lang]');
+
+    if (!innerElsWithLang.length) return currentLang || '';
+
+    const innerText = getTrimmedInnerText(node);
+
+    let innerTextLang = currentLang;
+
+    for (const el of node.childNodes) {
+      if (innerText === getTrimmedInnerText(el)) {
+        if (!(el instanceof HTMLElement || el instanceof SVGElement)) {
+          return currentLang || '';
+        }
+
+        const elLang = el.getAttribute('lang');
+        const childrenWithLang = el.querySelectorAll('[lang]');
+
+        if (!childrenWithLang.length) {
+          return elLang || currentLang || '';
+        } else {
+          // recursive call
+          return getLangOfInnerText(el, elLang || currentLang || '');
+        }
+      } else {
+        innerTextLang = '';
+      }
+    }
+
+    return innerTextLang || '';
+  }
+
   /** @type {Array<HTMLAnchorElement|SVGAElement>} */
   // @ts-expect-error - put into scope via stringification
   const anchorElements = getElementsInDocument('a'); // eslint-disable-line no-undef
 
+  // Check, if document has only one lang attribute in opening html tag. If so,
+  // there is no need to run the `getLangOfInnerText()` function with multiple
+  // possible DOM traversals
+  /** @type {Array<HTMLElement|SVGElement>} */
+  // @ts-expect-error - put into scope via stringification
+  const langElements = getElementsInDocument('[lang]'); // eslint-disable-line no-undef
+  const docHasOnlyHtmlLangAttr = (langElements.length === 1) &&
+    langElements[0].nodeName === 'HTML';
+  const lang = !docHasOnlyHtmlLangAttr ? null : langElements[0].getAttribute('lang');
```

Comment:
this needs to be normalized some how. Consider the valid language code `en-US`.

I don't recall how we could actually normalize to the base language code. In `shared/localization/locales.js` we form a mapping explicitly. I suppose here we could just split by `-` and return the first part to use as the loclae.

Reply:
I wouldn't normalize here. The gatherer just collects the lang code and passes it to the audit. If it should get normalized, then the last tool in the chain (the audit) should do that.

See also my response below to @adamraine about keeping the loop over language code segments.